src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,717 @@
+      addr_of_loop_write_set,
+      from_integer(assigns.size(), size_type()),
+      from_integer(0, size_type()),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      loop_head_location);
+    pre_loop_head_instrs.add(
+      goto_programt::make_function_call(call, loop_head_location));
+
+    pre_loop_head_instrs.destructive_append(assigns_instrs);
+  }
+
+  // goto HEAD;
+  pre_loop_head_instrs.add(
+    goto_programt::make_goto(loop_head, true_exprt{}, loop_head_location));
+
+  goto_function.body.destructive_insert(loop_head, pre_loop_head_instrs);
+  return history_var_map;
+}
+
+goto_programt::instructiont::targett
+dfcc_instrument_loopt::add_step_instructions(
+  const std::size_t loop_id,
+  const std::size_t cbmc_loop_id,
+  const irep_idt &function_id,
+  goto_functionst::goto_functiont &goto_function,
+  symbol_tablet &symbol_table,
+  goto_programt::targett loop_head,
+  goto_programt::targett loop_latch,
+  goto_programt &havoc_instrs,
+  const exprt &invariant,
+  const exprt::operandst &decreases_clauses,
+  const symbol_exprt &addr_of_loop_write_set,
+  const exprt &outer_write_set,
+  const symbol_exprt &initial_invariant,
+  const symbol_exprt &in_base_case,
+  const std::vector<symbol_exprt> &old_decreases_vars,
+  const irep_idt &language_mode)
+{
+  auto loop_head_location(loop_head->source_location());
+  dfcc_remove_loop_tags(loop_head_location);
+  goto_convertt converter(symbol_table, message_handler);
+
+  // ```
+  // STEP: // Loop step block: havoc the loop state
+  //   ASSERT(initial_invariant);
+  //   __write_set_check_inclusion(ws_loop, ws_parent);
+  //   in_base_case = false;
+  //   in_loop_havoc_block = true;
+  //   havoc(assigns_clause_targets);
+  //   in_loop_havoc_block = false;
+  //   ASSUME(loop_invariant);
+  //   old_variant = loop_decreases;
+  // ```
+
+  goto_programt step_instrs;
+  goto_programt::instructiont::targett step_case_target;
+
+  {
+    // We skip past it initially, because of the unconditional jump above,
+    // but jump back here if we get past the loop guard while in_base_case.
+    // `in_base_case = false;`
+    step_case_target = step_instrs.add(goto_programt::make_assignment(
+      in_base_case, false_exprt{}, loop_head_location));
+  }
+
+  {
+    // If we jump here, then the loop runs at least once,
+    // so add the base case assertion: `assert(initial_invariant)`
+    source_locationt check_location(loop_head_location);
+    check_location.set_property_class("loop_invariant_base");
+    check_location.set_comment(
+      "Check invariant before entry for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    code_assertt assertion{initial_invariant};
+    assertion.add_source_location() = check_location;
+    converter.goto_convert(assertion, step_instrs, language_mode);
```

Comment:
I don't think you need `goto_convert` here, `initial_invariant` is just a symbol expression, there won't be any cleanup needed.

Reply:
good point